File: dynamicProgramming/minCostClimbingStairs.java
package dynamicProgramming;

public class minCostClimbingStairs {
    public static void main(String[] args) {
        SolutionMCCS solution = new SolutionMCCS();
        System.out.println(solution.minCostClimbingStairs(new int[]{1, 100, 1, 1, 1, 100, 1, 1, 100, 1}));
    }
    
}


class SolutionMCCS {

    // Submitted by @Jiganesh 

    // TC : O(N)
    // SC : O(N)

    // Runtime: 8 ms, faster than 5.48% of Java online submissions for Min Cost Climbing Stairs.
    // Memory Usage: 44.7 MB, less than 5.17% of Java online submissions for Min Cost Climbing Stairs.
    public int minCostClimbingStairs(int[] cost) {
        
        
        int[] dp = new int[cost.length];
        dp[0]= cost[0];
        dp[1]= cost[1];
        
        for (int i=2 ; i<cost.length;i++){
            dp[i] = cost[i]+Math.min(dp[i-1], dp[i-2]);
        }
        
        int n = dp.length;
        return Math.min(dp[n-1], dp[n-2]);
    }

    // Efficient Solution
    
    // TC : O(N)
    // SC : O(1)

    // Runtime: 1 ms, faster than 89.31% of Java online submissions for Min Cost Climbing Stairs.
    // Memory Usage: 43.8 MB, less than 15.27% of Java online submissions for Min Cost Climbing Stairs.

    public int minCostClimbingStairsApproach2(int[] cost) {
        
        int a= cost[0];
        int b= cost[1];
        
        for (int i=2 ; i<cost.length;i++){
            int temp = b;
            b = cost[i]+Math.min(a, b);
            a = temp;
        }
        
        return Math.min(a,b);
    }
}

File: dynamicProgramming/minCostClimbingStairs.py
# https://leetcode.com/problems/min-cost-climbing-stairs/

class Solution(object):
    # Runtime: 27 ms, faster than 99.89% of Python online submissions for Min Cost Climbing Stairs.
    # Memory Usage: 13.6 MB, less than 55.39% of Python online submissions for Min Cost Climbing Stairs.
    
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """        
        dp = [float("inf")]*len(cost)
        dp[0] = cost[0]
        dp[1] = cost[1]
        
        
        for i in range (2, len(cost)):
            dp[i] = min(dp[i-1], dp[i-2])+cost[i]
            
        return min(dp[-1], dp[-2])
    
    # Space Optimizeds    
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """

        a = cost[0]
        b = cost[1]
        
        for i in range(2,len(cost)):
            a, b = b, cost[i]+min(a,b)
            
        return min(a, b)